### PR TITLE
feat/streaming ssr

### PR DESCRIPTION
**Pull request by GiveMe-A-Name**
# PR Details

Feat: Support React 18 Streaming SSR

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/renderToPipe.ts [hunk 1]
```diff
@@ -22,52 +17,56 @@ enum StreamingStatus {
...
+                  case StreamingStatus.Entry: {
+                    this.push(joinChunk(beforeEntry, chunk, afterEntry));
+                    streamingStaus = StreamingStatus.Streaming;
+                    break;
+                  }
+                  case StreamingStatus.Streaming: {
+                    this.push(chunk.toString());
+                    break;
+                  }
+                  case StreamingStatus.Leave: {
```
Comment (xuchaobei) [inline packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/renderToPipe.ts:47]: Leave 什么场景使用？

File: packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/template.ts [hunk 2]
```diff
@@ -43,14 +44,19 @@ export function createTemplates(
...
       fileContent?.split(HTML_SEPARATOR) || [];
+
+    // prepare inject objects.
     const loadableChunks =
       chunkExtractor?.getChunkAssets(chunkExtractor.chunks) || [];
     const loadableScripts = getLoadableScripts(chunkExtractor!);
     const styledComponentCSS = styleSheet?.getStyleTags() || '';
+
+    // templates injected some variables
     const builedBeforeTemplate = buildBeforeEntryTemplate(beforeEntryTemplate, {
```
Comment (xuchaobei) [inline packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/template.ts:60]: typo: builed

File: packages/server/prod-server/src/libs/render/cache/index.ts [hunk 3]
```diff
@@ -1,45 +1,78 @@
...
     if (cacheFile.isGarbage) {
-      const html = await renderFn(context);
-      const { cacheConfig } = context;
-
-      if (html && cacheConfig) {
-        await sprCache.set(cacheContext, html, cacheConfig);
-      }
-
-      return html;
+      const renderResult = await renderFn(context);
```
Comment (xuchaobei) [inline packages/server/prod-server/src/libs/render/cache/index.ts:56]: 和 56-62 行重复了，封装成一个函数？

File: packages/server/prod-server/src/server/modern-server.ts [hunk 4]
```diff
@@ -492,9 +494,18 @@ export class ModernServer implements ModernServerInterface {
       res.setHeader('content-type', file.contentType);
       res.end(response);
     } else {
-      const { pipe } = file;
       res.setHeader('content-type', file.contentType);
-      pipe?.(res);
+      response
+        .pipe(
```
Comment (xuchaobei) [inline packages/server/prod-server/src/server/modern-server.ts:477]: 如果是使用真正的流 （onShellReady 中 pipe)，每次有新的流产生时，都会经过这个 Transform 做转换，会导致数据重复吧？

File: packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/renderToPipe.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import { Transform, Writable } from 'stream';
+import {
+  RenderToPipeableStreamOptions,
+  renderToPipeableStream,
+} from 'react-dom/server';
+import { InjectTemplate } from './type';
+
+export type Pipe<T extends Writable> = (output: T) => Promise<T>;
+
+enum StreamingStatus {
```
Comment (xuchaobei) [inline packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/renderToPipe.ts:10]: 这几个状态可以做下注释，命名上是否能再优化下，比如 entry 改成 start ? entry 和 streaming 不是同类型的词语

File: packages/runtime/plugin-runtime/src/ssr/serverRender/utils.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  const check = (scripts: string) =>
+    (scripts || '').includes('__LOADABLE_REQUIRED_CHUNKS___ext');
+
+  const scripts = extractor.getScriptTags();
+
+  if (!check(scripts)) {
+    return '';
+  }
+
+  return (
```
Comment (xuchaobei) [inline packages/runtime/plugin-runtime/src/ssr/serverRender/utils.ts:13]: 这段逻辑的作用是什么呢？

**Review by xuchaobei — CHANGES_REQUESTED** [review]

File: packages/server/prod-server/src/libs/render/cache/index.ts (hunk 3 above)
Comment (GiveMe-A-Name) [reply packages/server/prod-server/src/libs/render/cache/index.ts:58]: done

File: packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/template.ts (hunk 2 above)
Comment (GiveMe-A-Name) [reply packages/runtime/plugin-runtime/src/ssr/serverRender/renderToStream/template.ts]: done

**Review by xuchaobei — APPROVED** [review]

**Pull request merged**